File: avalon/composition/impl/src/java/org/apache/avalon/composition/model/impl/DefaultContextModel.java
/* 
 * Copyright 2004 Apache Software Foundation
 * Licensed  under the  Apache License,  Version 2.0  (the "License");
 * you may not use  this file  except in  compliance with the License.
 * You may obtain a copy of the License at 
 * 
 *   http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed  under the  License is distributed on an "AS IS" BASIS,
 * WITHOUT  WARRANTIES OR CONDITIONS  OF ANY KIND, either  express  or
 * implied.
 * 
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.avalon.composition.model.impl;

import java.util.Map;
import java.util.Hashtable;
import java.lang.reflect.Constructor;

import org.apache.avalon.composition.data.ContextDirective;
import org.apache.avalon.composition.data.EntryDirective;
import org.apache.avalon.composition.data.ImportDirective;
import org.apache.avalon.composition.data.ConstructorDirective;
import org.apache.avalon.composition.model.EntryModel;
import org.apache.avalon.composition.model.ContextModel;
import org.apache.avalon.composition.model.ComponentModel;
import org.apache.avalon.composition.model.ContainmentModel;
import org.apache.avalon.composition.model.ModelException;
import org.apache.avalon.composition.provider.ComponentContext;

import org.apache.avalon.excalibur.i18n.ResourceManager;
import org.apache.avalon.excalibur.i18n.Resources;

import org.apache.avalon.framework.context.Context;
import org.apache.avalon.framework.context.ContextException;
import org.apache.avalon.framework.logger.AbstractLogEnabled;
import org.apache.avalon.framework.logger.Logger;

import org.apache.avalon.meta.info.ContextDescriptor;
import org.apache.avalon.meta.info.EntryDescriptor;


/**
 * <p>Specification of a context model from which a 
 * a fully qualifed context can be established.</p>
 *
 * @author <a href="mailto:dev@avalon.apache.org">Avalon Development Team</a>
 * @version $Revision: 1.12 $ $Date: 2004/02/22 17:28:30 $
 */
public class DefaultContextModel extends DefaultDependent implements ContextModel
{
    //==============================================================
    // static
    //==============================================================

    private static final Resources REZ =
      ResourceManager.getPackageResources( DefaultContextModel.class );

    /**
     * The default context implementation class to be used if
     * no context class is defined.
     */
    public static final Class DEFAULT_CONTEXT_CLASS = 
      DefaultContext.class;

    public static boolean isaStandardKey( String key )
    {
        return ( key.startsWith( "urn:avalon:" ) 
          || key.startsWith( "urn:composition:" ));
    }

    //==============================================================
    // immutable state
    //==============================================================

    private final ContextDescriptor m_descriptor;

    private final ContextDirective m_directive;

    private final ComponentContext m_context;

    private final Class m_strategy;

    private final Map m_models = new Hashtable();

    private final Map m_map = new Hashtable();

    private final Context m_componentContext;

    //==============================================================
    // constructor
    //==============================================================

   /**
    * <p>Default implementation of the context model.  The implementation
    * takes an inital system context as the base for context value 
    * establishment and uses this to set standard context entries.</p>
    *
    * @param logger the logging channel
    * @param descriptor the contextualization stage descriptor that describes
    *   the set of context entries that the component type is requesting
    * @param directive the contextualization directive that describes a set 
    *   of context entry creation strategies
    * @param context the component model context argument
    */
    public DefaultContextModel( 
      Logger logger, ContextDescriptor descriptor, 
      ContextDirective directive, ComponentContext context )
      throws ModelException
    {
        super( logger );

        if( null == descriptor )
        {
            throw new NullPointerException( "descriptor" );
        }

        if( null == context )
        {
            throw new NullPointerException( "context" );
        }

        m_descriptor = descriptor;
        m_directive = directive;
        m_context = context;

        ClassLoader classLoader = context.getClassLoader();
        m_strategy = loadStrategyClass( descriptor, classLoader );

        //
        // get the set of context entries declared by the component type
        // and for for each entry determine the context entry model to 
        // use for context entry value resolution
        //

        EntryDescriptor[] entries = descriptor.getEntries();
        for( int i=0; i<entries.length; i++ )
        {
            EntryDescriptor entry = entries[i];
            String alias = entry.getAlias();
            final String key = entry.getKey();
            if( isaStandardKey( key ) )
            {
                 DefaultImportModel model = 
                   new DefaultImportModel( entry, key, m_context );
                 setEntryModel( alias, model );
            }
            else
            {
                //
                // its a non standard context entry so check for a 
                // entry directive with a matching key to define
                // the mechanism for building the context entry
                //

                EntryDirective entryDirective = 
                  directive.getEntryDirective( key );
                if( null == entryDirective )
                {
                    if( entry.isRequired() )
                    {
                        final String error = 
                          REZ.getString( 
                            "context.missing-directive.error", key );
                        throw new ModelException( error );
                    }
                }
                else
                {
                    //
                    // there are only two context entry models - import
                    // and constructor - identify the model to use then add
                    // the resolved model to the map
                    //

                    if( entryDirective instanceof ImportDirective )
                    {
                        //
                        // importing under an alias of a container scoped key
                        //

                        ImportDirective importDirective = 
                          (ImportDirective) entryDirective;
                        String ref = importDirective.getImportKey();
                        DefaultImportModel model = 
                          new DefaultImportModel( entry, ref, m_context );
                        setEntryModel( alias, model );
                    }
                    else if( entryDirective instanceof ConstructorDirective )
                    {
                        ConstructorDirective constructor = 
                          (ConstructorDirective) entryDirective;
                        DefaultConstructorModel model = 
                          new DefaultConstructorModel( 
                            entry, 
                            constructor, 
                            context, 
                            m_map );
                        setEntryModel( alias, model );
                    }
                    else
                    {
                        String modelClass = 
                          entryDirective.getClass().getName();
                        final String error = 
                          REZ.getString( 
                            "context.unsupported-directive.error", 
                            key, modelClass );
                        throw new ModelException( error );
                    }
                }
            }
        }

        m_componentContext = 
          createComponentContext( classLoader, descriptor, directive, m_map );
    }

    
    //==============================================================
    // ContextModel
    //==============================================================

   /**
    * Return the set of entry models associated with this context model.
    * 
    * @return the entry models
    */
    public EntryModel[] getEntryModels()
    {
        return (EntryModel[]) m_map.values().toArray( new EntryModel[0] );
    }

   /**
    * Return an entry model matching the supplied key.
    * 
    * @return the entry model or null if tyhe key is unknown
    */
    public EntryModel getEntryModel( String key )
    {
        return (EntryModel) m_map.get( key ); 
    }

   /**
    * Set the entry model relative to a supplied key.
    * 
    * @param key the entry key
    * @param model the entry model
    */
    public void setEntryModel( String key, EntryModel model )
    {
        m_map.put( key, model ); 
    }

   /**
    * Set the entry to a suplied value.
    * 
    * @param key the entry key
    * @param value the entry value
    */
    public void setEntry( String key, Object value )
    {
        EntryDescriptor descriptor = m_descriptor.getEntry( key );
        OverrideEntryModel model = 
          new OverrideEntryModel( descriptor, value );
        setEntryModel( key, model );
    }

   /**
    * Return the class representing the contextualization stage interface.
    * 
    * @return the class representing the contextualization interface
    */
    public Class getStrategyClass()
    {
        return m_strategy;
    }

   /**
    * Return the context object established for the component.
    * 
    * @return the context object
    */
    public Context getContext()
    {
        return m_componentContext;
    }

    //==============================================================
    // implementation
    //==============================================================

   /**
    * Load the contextualization strategy class.
    * @param descriptor the context descriptor
    * @param classloader the classloader 
    * @return the strategy class
    */
    private Class loadStrategyClass( 
       ContextDescriptor descriptor, ClassLoader classloader )
      throws ModelException
    {
        final String strategy = m_descriptor.getAttribute( 
          ContextDescriptor.STRATEGY_KEY, null );
        if( strategy != null )
        {
            try
            {
                Class clazz = classloader.loadClass( strategy );
                if( getLogger().isDebugEnabled() )
                {
                    final String message = 
                      REZ.getString( "context.strategy.custom", strategy );
                    getLogger().debug( message );
                }
                return clazz;
            }
            catch( ClassNotFoundException e )
            {
                final String error = 
                  REZ.getString( "context.strategy.custom.missing.error", strategy );
                throw new ModelException( error );
            }
            catch( Throwable e )
            {
                final String error = 
                  REZ.getString( "context.strategy.custom.unexpected.error", strategy );
                throw new ModelException( error, e );
            }
        }
        else
        {
            try
            {
                Class clazz = classloader.loadClass( DEFAULT_STRATEGY_CLASSNAME );
                if( getLogger().isDebugEnabled() )
                {
                    final String message = 
                      REZ.getString( "context.strategy.avalon" );
                    getLogger().debug( message );
                }
                return clazz;
            }
            catch( ClassNotFoundException e )
            {
                final String error = 
                  REZ.getString( 
                    "context.strategy.avalon.missing.error", 
                    DEFAULT_STRATEGY_CLASSNAME );
                throw new ModelException( error );
            }
            catch( Throwable e )
            {
                final String error = 
                  REZ.getString( 
                    "context.strategy.avalon.unexpected.error",
                    DEFAULT_STRATEGY_CLASSNAME );
                throw new ModelException( error, e );
            }
        }
    }

   /**
    * Creates a component context instance.
    * 
    * @param classloader the deployment context classloader
    * @param descriptor the context descriptor
    * @param directive the context directive
    * @return the context object compliant with the context casting
    *   constraints declared by the component type
    * @exception ModelException if an error occurs while attempting to 
    *   construct the context instance
    */
    private Context createComponentContext( 
      ClassLoader classloader, 
      ContextDescriptor descriptor, 
      ContextDirective directive, Map map )
      throws ModelException
    {
        Class clazz = loadContextClass( directive, classloader );
        validateCastingConstraint( descriptor, classloader, clazz );
        Context base = new DefaultContext( map );

        if( clazz.equals( DefaultContext.class ) ) return base; 

        //
        // its a custom context object so we need to create it 
        // using the classic context object as the constructor 
        // argument
        //

        try
        {
            Constructor constructor = clazz.getConstructor(
                new Class[]{ Context.class } );
            return (Context) constructor.newInstance( new Object[]{ base } );
        }
        catch( NoSuchMethodException e )
        {
            final String error =
              REZ.getString( "context.non-compliance-constructor.error", clazz.getName() );
            throw new ModelException( error, e );
        }
        catch( Throwable e )
        {
            final String error =
              REZ.getString( "context.custom-unexpected.error", clazz.getName() );
            throw new ModelException( error, e );
        }
    }

   /**
    * Load the context implementation class.
    * @param directive the context directive (possibly null)
    * @param classLoader the classloader 
    * @return the strategy class
    */
    private Class loadContextClass( 
       ContextDirective directive, ClassLoader classLoader )
      throws ModelException
    {
        if( directive == null )
        {
            return DEFAULT_CONTEXT_CLASS;
        }

        final String classname = directive.getClassname();
        if( classname == null )
        {
            return DEFAULT_CONTEXT_CLASS;
        }
        else
        {
            try
            {
                return classLoader.loadClass( classname );
            }
            catch( Throwable e )
            {
                final String error = 
                  "Cannot load custom context implementation class: "
                  + classname;
                throw new ModelException( error, e );
            }
        }
    }

   /**
    * Validate that the context implememtation class implements
    * any casting constraint declared or implied by the context 
    * descriptor.
    * 
    * @param descriptor the context descriptor
    * @param classLoader the classloader
    * @param clazz the context implementation class
    * @exception if a validation failure occurs
    */
    private void validateCastingConstraint( 
      ContextDescriptor descriptor, ClassLoader classLoader, Class clazz )
      throws ModelException
    {

        Class castingClass = null;

        final String castingClassName = 
          descriptor.getContextInterfaceClassname();

        if( castingClassName != null )
        {
            try
            {
                castingClass =
                  classLoader.loadClass( castingClassName );
            }
            catch( Throwable e )
            {
                final String error = 
                  "Cannot load custom context interface class: "
                  + castingClassName;
                throw new ModelException( error, e );
            }
        }
        else
        {
            try
            {
                castingClass =
                  classLoader.loadClass( Context.class.getName() );
            }
            catch( Throwable e )
            {
                final String error = 
                  "Cannot load standard Avalon context interface class: "
                  + Context.class.getName();
                throw new ModelException( error, e );
            }
        }

        if( !castingClass.isAssignableFrom( clazz ) )
        {
            final String error = 
              "Supplied context implementation class: " 
              + clazz.getName() 
              + " does not implement the interface: " 
              + castingClass.getName()
              + ".";
            throw new ModelException( error );
        }
    }
}

File: legacy/avalon-sandbox/avalon-net/Composition/ModelDefault/DefaultContextModel.cs
/* 
* Copyright 2003-2004 The Apache Software Foundation
* Licensed  under the  Apache License,  Version 2.0  (the "License");
* you may not use  this file  except in  compliance with the License.
* You may obtain a copy of the License at 
* 
*   http://www.apache.org/licenses/LICENSE-2.0
* 
* Unless required by applicable law or agreed to in writing, software
* distributed  under the  License is distributed on an "AS IS" BASIS,
* WITHOUT  WARRANTIES OR CONDITIONS  OF ANY KIND, either  express  or
* implied.
* 
* See the License for the specific language governing permissions and
* limitations under the License.
*/
namespace Apache.Avalon.Composition.Model.Default
{
	using System;

	using Apache.Avalon.Framework;
	using Apache.Avalon.Composition.Data;
	using Apache.Avalon.Composition.Model;
	using Apache.Avalon.Meta;
	
	/// <summary> <p>Specification of a context model from which a 
	/// a fully qualifed context can be established.</p>
	/// 
	/// </summary>
	/// <author>  <a href="mailto:dev@avalon.apache.org">Avalon Development Team</a>
	/// </author>
	/// <version>  $Revision: 1.2 $ $Date: 2004/02/29 18:07:17 $
	/// </version>
	public class DefaultContextModel : DefaultDependent, IContextModel
	{
		private void  InitBlock()
		{
			m_models = new System.Collections.Hashtable();
			m_map = new System.Collections.Hashtable();
		}

		//==============================================================
		// static
		//==============================================================
		
		/// <summary> The default context implementation class to be used if
		/// no context class is defined.
		/// </summary>
		//UPGRADE_NOTE: Final was removed from the declaration of 'DEFAULT_CONTEXT_CLASS '. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1003"'
		public static readonly System.Type DEFAULT_CONTEXT_CLASS = typeof(DefaultContext);
		
		//==============================================================
		// immutable state
		//==============================================================
		
		//UPGRADE_NOTE: Final was removed from the declaration of 'm_descriptor '. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1003"'
		private ContextDescriptor m_descriptor;
		
		//UPGRADE_NOTE: Final was removed from the declaration of 'm_directive '. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1003"'
		private ContextDirective m_directive;
		
		//UPGRADE_NOTE: Final was removed from the declaration of 'm_context '. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1003"'
		private IComponentContext m_context;
		
		//UPGRADE_NOTE: Final was removed from the declaration of 'm_strategy '. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1003"'
		private System.Type m_strategy;
		
		//UPGRADE_NOTE: Final was removed from the declaration of 'm_models '. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1003"'
		//UPGRADE_NOTE: The initialization of  'm_models' was moved to method 'InitBlock'. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1005"'
		private System.Collections.IDictionary m_models;
		
		//UPGRADE_NOTE: Final was removed from the declaration of 'm_map '. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1003"'
		//UPGRADE_NOTE: The initialization of  'm_map' was moved to method 'InitBlock'. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1005"'
		private System.Collections.IDictionary m_map;
		
		//UPGRADE_NOTE: Final was removed from the declaration of 'm_componentContext '. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1003"'
		private IContext m_componentContext;
		
		//==============================================================
		// constructor
		//==============================================================
		
		/// <summary> <p>Default implementation of the context model.  The implementation
		/// takes an inital system context as the base for context value 
		/// establishment and uses this to set standard context entries.</p>
		/// 
		/// </summary>
		/// <param name="logger">the logging channel
		/// </param>
		/// <param name="descriptor">the contextualization stage descriptor
		/// </param>
		/// <param name="directive">the contextualization directive
		/// </param>
		/// <param name="context">the deployment context
		/// </param>
		public DefaultContextModel(ILogger logger, ContextDescriptor descriptor, ContextDirective directive, IComponentContext context):base(logger)
		{
			InitBlock();
			
			if (null == descriptor)
			{
				throw new System.ArgumentNullException("descriptor");
			}
			
			if (null == context)
			{
				throw new System.ArgumentNullException("context");
			}
			
			m_descriptor = descriptor;
			m_directive = directive;
			m_context = context;
			
			//UPGRADE_ISSUE: Class 'java.lang.ClassLoader' was not converted. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1000_javalangClassLoader"'
			m_strategy = loadStrategyClass(descriptor);
			
			//
			// get the set of context entries declared by the component type
			// and for for each entry determine the context entry model to 
			// use for context entry value resolution
			//
			
			EntryDescriptor[] entries = descriptor.Entries;
			for (int i = 0; i < entries.Length; i++)
			{
				EntryDescriptor entry = entries[i];
				//UPGRADE_NOTE: Final was removed from the declaration of 'key '. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1003"'
				String key = entry.Key;
				if (key.StartsWith("urn:avalon:"))
				{
					try
					{
						System.Object value_Renamed = m_context.Resolve(key);
						object tempObject;
						tempObject = value_Renamed;
						m_map[key] = tempObject;
						System.Object generatedAux = tempObject;
					}
					catch (ContextException )
					{
						if (entry.Required)
						{
							String error = "context.non-standard-avalon-key.error" + " " + key;
							throw new ModelException(error);
						}
					}
				}
				else if (key.Equals(Apache.Avalon.Composition.Model.IContainmentModel_Fields.KEY))
				{
					//
					// TODO: check that the component has permission
					// to access the containment model
					//
					
					object tempObject2;
					tempObject2 = context.ContainmentModel;
					m_map[Apache.Avalon.Composition.Model.IContainmentModel_Fields.KEY] = tempObject2;
					System.Object generatedAux2 = tempObject2;
				}
				else if (key.StartsWith("urn:composition:"))
				{
					try
					{
						System.Object value_Renamed = m_context.SystemContext[key];
						object tempObject3;
						tempObject3 = value_Renamed;
						m_map[key] = tempObject3;
						System.Object generatedAux3 = tempObject3;
					}
					catch (ContextException )
					{
						if (entry.Required)
						{
							String error = "context.non-standard-avalon-key.error" + " " + key;
							throw new ModelException(error);
						}
					}
				}
				else
				{
					//
					// its a non standard context entry so check for a 
					// entry directive with a matching key to define
					// the mechanism for building the context entry
					//
					
					EntryDirective entryDirective = directive.getEntryDirective(key);
					if (null == entryDirective)
					{
						if (entry.Required)
						{
							String error = "context.missing-directive.error" + " " + key;
							throw new ModelException(error);
						}
					}
					else
					{
						//
						// there are only two context entry models - import
						// and constructor - identify the model to use then add
						// the resolved model to the map
						//
						
						if (entryDirective is ImportDirective)
						{
							ImportDirective importDirective = (ImportDirective) entryDirective;
							DefaultImportModel model = new DefaultImportModel(entry, importDirective, context, m_map);
							m_context.Register(model);
							object tempObject4;
							tempObject4 = model.Value;
							m_map[key] = tempObject4;
							System.Object generatedAux4 = tempObject4;
						}
						else if (entryDirective is ConstructorDirective)
						{
							ConstructorDirective constructor = (ConstructorDirective) entryDirective;
							DefaultConstructorModel model = new DefaultConstructorModel(entry, constructor, context, m_map);
							m_context.Register(model);
							object tempObject5;
							tempObject5 = model.Value;
							m_map[key] = tempObject5;
							System.Object generatedAux5 = tempObject5;
						}
						else
						{
							String modelClass = entryDirective.GetType().FullName;
							String error = "context.unsupported-directive.error" + " " + key + " " + modelClass;
							throw new ModelException(error);
						}
					}
				}
			}
			
			m_componentContext = CreateComponentContext(m_context, descriptor, directive);
			
			if (Logger.IsDebugEnabled)
			{
				Logger.Debug("context: " + m_map);
			}
		}
		
		//==============================================================
		// IContextModel
		//==============================================================
		
		/// <summary> Return the context object established for the component.
		/// 
		/// </summary>
		/// <returns> the context object
		/// </returns>
		public virtual IContext Context
		{
			get
			{
				return m_componentContext;
			}
			
		}

		/// <summary> Return the class representing the contextualization stage interface.
		/// 
		/// </summary>
		/// <returns> the class representing the contextualization interface
		/// </returns>
		public virtual System.Type StrategyClass
		{
			get
			{
				return m_strategy;
			}
		}
		
		//==============================================================
		// implementation
		//==============================================================
		
		/// <summary> Load the contextualization strategy class.</summary>
		/// <param name="descriptor">the context descriptor
		/// </param>
		/// <param name="classloader">the classloader 
		/// </param>
		/// <returns> the strategy class
		/// </returns>
		private System.Type loadStrategyClass(ContextDescriptor descriptor)
		{
			//UPGRADE_NOTE: Final was removed from the declaration of 'strategy '. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1003"'
			String strategy = m_descriptor.GetAttribute(ContextDescriptor.STRATEGY_KEY, null);
			if ((System.Object) strategy != null)
			{
				try
				{
					//UPGRADE_ISSUE: Method 'java.lang.ClassLoader.loadClass' was not converted. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1000_javalangClassLoader"'
					System.Type type = Type.GetType(strategy);
					if (Logger.IsDebugEnabled)
					{
						//UPGRADE_NOTE: Final was removed from the declaration of 'message '. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1003"'
						String message = "context.strategy.custom" + " " + strategy;
						Logger.Debug(message);
					}
					return type;
				}
				//UPGRADE_NOTE: Exception 'java.lang.ClassNotFoundException' was converted to 'System.Exception' which has different behavior. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1100"'
				catch (System.Exception )
				{
					//UPGRADE_NOTE: Final was removed from the declaration of 'error '. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1003"'
					String error = "context.strategy.custom.missing.error" + " " + strategy;
					throw new ModelException(error);
				}
			}
			else
			{
				try
				{
					System.Type type = Type.GetType(Apache.Avalon.Composition.Model.IContextModel_Fields.DEFAULT_STRATEGY_CLASSNAME);
					if (Logger.IsDebugEnabled)
					{
						String message = "context.strategy.avalon";
						Logger.Debug(message);
					}
					return type;
				}
				catch (System.Exception )
				{
					String error = "context.strategy.avalon.missing.error" + " " + Apache.Avalon.Composition.Model.IContextModel_Fields.DEFAULT_STRATEGY_CLASSNAME;
					throw new ModelException(error);
				}
			}
		}
		
		/// <summary> Creates a compoent context using a deployment context that 
		/// has been pre-populated with constom context entry models.
		/// 
		/// </summary>
		/// <param name="context">the deployment context
		/// </param>
		/// <param name="descriptor">the context descriptor
		/// </param>
		/// <param name="directive">the context directive
		/// </param>
		/// <returns> the context object compliant with the context casting
		/// constraints declared by the component type
		/// </returns>
		/// <exception cref=""> ModelException if an error occurs while attempting to 
		/// construct the context instance
		/// </exception>
		private IContext CreateComponentContext(IComponentContext context, ContextDescriptor descriptor, ContextDirective directive)
		{
			//UPGRADE_ISSUE: Class 'java.lang.ClassLoader' was not converted. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1000_javalangClassLoader"'
			System.Type type = loadContextClass(directive);
			validateCastingConstraint(descriptor, type);
			IContext base_Renamed = new DefaultContext(context);
			
			if (type.Equals(typeof(DefaultContext)))
				return base_Renamed;
			
			//
			// its a custom context object so we need to create it 
			// using the classic context object as the constructor 
			// argument
			//
			
			try
			{
				System.Reflection.ConstructorInfo constructor = type.GetConstructor(new System.Type[]{typeof(IContext)});
				//UPGRADE_ISSUE: Method 'java.lang.reflect.Constructor.newInstance' was not converted. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1000_javalangreflectConstructornewInstance_javalangObject[]"'
				return (IContext) constructor.Invoke(new System.Object[]{base_Renamed});
			}
			catch (System.MethodAccessException e)
			{
				//UPGRADE_NOTE: Final was removed from the declaration of 'error '. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1003"'
				String error = "context.non-compliance-constructor.error" + " " + type.FullName;
				throw new ModelException(error, e);
			}
			catch (System.Exception e)
			{
				//UPGRADE_NOTE: Final was removed from the declaration of 'error '. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1003"'
				String error = "context.custom-unexpected.error" + " " + type.FullName;
				throw new ModelException(error, e);
			}
		}
		
		/// <summary> Load the context implementation class.</summary>
		/// <param name="directive">the context directive (possibly null)
		/// </param>
		/// <param name="classLoader">the classloader 
		/// </param>
		/// <returns> the strategy class
		/// </returns>
		//UPGRADE_ISSUE: Class 'java.lang.ClassLoader' was not converted. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1000_javalangClassLoader"'
		private System.Type loadContextClass(ContextDirective directive)
		{
			if (directive == null)
			{
				return DEFAULT_CONTEXT_CLASS;
			}
			
			String typename = directive.Classname;
			if ((System.Object) typename == null)
			{
				return DEFAULT_CONTEXT_CLASS;
			}
			else
			{
				try
				{
					return Type.GetType(typename);
				}
				catch (System.Exception e)
				{
					String error = "Cannot load custom context implementation class: " + typename;
					throw new ModelException(error, e);
				}
			}
		}
		
		/// <summary> Validate that the context implememtation class implements
		/// any casting constraint declared or implied by the context 
		/// descriptor.
		/// 
		/// </summary>
		/// <param name="descriptor">the context descriptor
		/// </param>
		/// <param name="classLoader">the classloader
		/// </param>
		/// <param name="type">the context implementation class
		/// </param>
		/// <exception cref=""> if a validation failure occurs
		/// </exception>
		private void  validateCastingConstraint(ContextDescriptor descriptor, System.Type type)
		{
			System.Type castingClass = descriptor.ContextInterface;
			
			if (castingClass == null)
			{
				try
				{
					castingClass = typeof(IContext);
				}
				catch (System.Exception e)
				{
					String error = "Cannot load standard Avalon context interface class: " + typeof(IContext).FullName;
					throw new ModelException(error, e);
				}
			}
			
			if (!castingClass.IsAssignableFrom(type))
			{
				String error = "Supplied context implementation class: " + type.FullName + " does not implement the interface: " + castingClass.FullName + ".";
				throw new ModelException(error);
			}
		}
	}
}